### PR TITLE
[Feature] Unify the cuda stream used in core library

### PR DESCRIPTION
**Pull request by chang-l**
## Description
To fix https://github.com/dmlc/dgl/issues/4471, I changed all hard-coded default stream to DGL internal stream.

- Enforced `CopyDataFromTo` and its associates to use stream `thr_entry->stream`
- Fix all `CUDA_KERNEL_CALL` to use `thr_entry->stream`
- Fix cub submodule to use `thr_entry->stream`
- Initialized dgl stream to `torch.cuda.current_stream` for each thread, to ensure backward propag. still uses the right stream.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Tests
- Initial test: 
  With the script from https://github.com/dmlc/dgl/issues/4436, run on a side stream and it will not crash anymore.
- [GCN example](https://github.com/dmlc/dgl/blob/master/examples/pytorch/gcn/train.py)
   Switch to a non-default stream (`torch.cuda.set_stream()`), then align DGL stream using `with dgl._ffi.streams.stream(s)`
   Results are consistent and profiler shows 100% kernels launch on the specified side stream.

- [Graphsage example](https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/node_classification.py)
   Switch to a non-default stream (`torch.cuda.set_stream()`), then align DGL stream using `with dgl._ffi.streams.stream(s)`
   Results are consistent and profiler shows ~70% kernels launch on the specified side stream, other (non-default) streams are all for non-blocking data-transfer in layer-wise inference stage (issued from pytorch). Cross-checked/confirmed with the reference profiling from default stream results.

- [GAT example](https://github.com/dmlc/dgl/blob/master/examples/pytorch/gat/train.py)
   Switch to a non-default stream (`torch.cuda.set_stream()`), then align DGL stream using `with dgl._ffi.streams.stream(s)`
   Results are consistent and profiler shows ~99.5% kernels launch on the specified side stream, the rest (<0.5%) are from `cudaMemcpy` issued on the legacy default stream (auto-sync blocking stream). These `cudaMemcpy` functions are ALL called from the following function:
https://github.com/dmlc/dgl/blob/2dd06b21529d77fa398706444f6db080378e8a79/src/array/cuda/macro.cuh#L16-L17

- [Heter-RGCN example](https://github.com/dmlc/dgl/blob/master/examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py)
   Switch to a non-default stream in pytorch, then align to the same stream for DGL.
   Results are consistent and profiler shows 100% kernels launch on the specified side stream.

- Multi-GPU runs ([graphsage](https://github.com/dmlc/dgl/blob/master/examples/pytorch/multigpu/multi_gpu_node_classification.py), and [rgcn](https://github.com/dmlc/dgl/blob/master/examples/pytorch/rgcn/entity_sample_multi_gpu.py))
   For multi-gpu runs, some un-specified stream launches are from (1) NCCL and (2) pytorch.

### CODE REVIEW

File: include/dgl/runtime/ndarray.h [hunk 1]
```diff
@@ -147,39 +147,42 @@ class NDArray {
...
-   */
-  inline void CopyTo(DLTensor *other,
-                     const DGLStreamHandle &stream = nullptr) const;
-  inline void CopyTo(const NDArray &other,
-                     const DGLStreamHandle &stream = nullptr) const;
+   * \param stream The stream to perform the copy context,
+   * \note The copy may happen synchronously if specified stream = dgl internal stream.
+   *       Otherwise, synchronization is necessary.
+   */
+  inline void CopyFromAsync(DLTensor* other,
```
Comment (yaox12) [inline include/dgl/runtime/ndarray.h:168]: I doubt if we really need `CopyFromAsync`/`CopyToAsync`. Where will they be used?

File: src/array/cuda/array_index_select.cu [hunk 2]
```diff
@@ -80,7 +81,7 @@ DType IndexSelect(NDArray array, int64_t index) {
   device->CopyDataFromTo(
       static_cast<DType*>(array->data) + index, 0, reinterpret_cast<DType*>(&ret), 0,
       sizeof(DType), array->ctx, DLContext{kDLCPU, 0},
-      array->dtype, nullptr);
+      array->dtype, thr_entry->stream);
```
Comment (yaox12) [inline src/array/cuda/array_index_select.cu:84]: How about omitting stream here?

File: src/dataloading/async_transferer.cc [hunk 3]
```diff
@@ -72,7 +72,7 @@ TransferId AsyncTransferer::StartTransfer(
     // get tensor information
     t.event.reset(new Event);
     CUDA_CALL(cudaEventCreate(&t.event->id));
-    t.dst.CopyFrom(t.src, stream_);
+    t.dst.CopyFromAsync(t.src, stream_);
```
Comment (yaox12) [inline src/dataloading/async_transferer.cc:75]: I think this is a legacy file and not used anymore. The corresponding Python code is under `dgl._dataloading`. Maybe we should remove this file in v1.0. cc @nv-dlasalle @mufeili

File: src/array/cuda/disjoint_union.cu [hunk 4]
```diff
@@ -135,17 +135,17 @@ COOMatrix DisjointUnionCoo(const std::vector<COOMatrix>& coos) {
   device->CopyDataFromTo(
       &prefix_elm[coos.size()], 0, &n_elements, 0,
       sizeof(IdType), coos[0].row->ctx, DGLContext{kDLCPU, 0},
-      coos[0].row->dtype, 0);
+      coos[0].row->dtype, stream);
```
Comment (yaox12) [inline src/array/cuda/disjoint_union.cu:138]: And here.

File: src/array/cuda/disjoint_union.cu [hunk 5]
```diff
@@ -84,7 +84,7 @@ void _Merge(IdType** arrs, IdType* prefix, IdType* offset, IdType* out,
 
   device->CopyDataFromTo(
       arrs, 0, arrs_dev, 0, sizeof(IdType*)*n_arrs,
-      DGLContext{kDLCPU, 0}, ctx, dtype, 0);
+      DGLContext{kDLCPU, 0}, ctx, dtype, stream);
```
Comment (yaox12) [inline src/array/cuda/disjoint_union.cu:87]: Same. And probably remove `cudaStream_t stream` from arguments.

File: src/array/cuda/negative_sampling.cu [hunk 6]
```diff
@@ -174,7 +174,7 @@ std::pair<IdArray, IdArray> CSRGlobalUniformNegativeSampling(
   CUDA_CALL(cub::DeviceSelect::If(
         tmp, tmp_size, begin, out_begin, num_out_cuda, num_actual_samples, op,
         thr_entry->stream));
-  num_out = cuda::GetCUDAScalar(device, ctx, num_out_cuda, static_cast<cudaStream_t>(0));
+  num_out = cuda::GetCUDAScalar(device, ctx, num_out_cuda, thr_entry->stream);
```
Comment (yaox12) [inline src/array/cuda/negative_sampling.cu:177]: Consider removing `stream` from `cuda::GetCUDAScalar()`.

File: include/dgl/runtime/ndarray.h (hunk 1 above)
Comment (chang-l) [reply include/dgl/runtime/ndarray.h:168]: Not really. This is mostly for future I guess. Maybe I can comment/mark it as `Todo`...

File: src/array/cuda/array_index_select.cu (hunk 2 above)
Comment (chang-l) [reply src/array/cuda/array_index_select.cu:84]: Sure. I will get rid of the extra `thr_entry->stream`

File: src/array/cuda/disjoint_union.cu (hunk 5 above)
Comment (chang-l) [reply src/array/cuda/disjoint_union.cu:87]: Sure.

**Comment by chang-l** [comment]
Once I resolve the above comments, there are still two remaining issues for this PR, according to my profiling:

(1) This PR depends on Xin's PR: https://github.com/dmlc/dgl/pull/4472, to completely unify the dgl internal stream.
(2) In pytorch, backward propag. uses a different thread, which makes the dgl stream `CUDAThreadEntry->ThreadLocal()->stream` out of control within `backward propg` section, i.e., it will always use the default stream for a different thread. 
I am thinking maybe we should avoid declaring cuda stream as a _thread-local_ variable, because cuda stream itself is thread-safe...

Please let me know your thoughts @yaox12 @nv-dlasalle @BarclayII @TristonC

**Comment by yaox12** [comment]
Yesterday I have a brief sync with @jermainewang and he suggests we always get the current stream from PyTorch (via TensorAdaptor maybe) instead of `CUDAThreadEntry->ThreadLocal()->stream`. How do you think about it? @chang-l 

Besides, I think PyTorch also stores the current CUDA stream in a thread-local variable. How does it handle the backward thread?
https://github.com/pytorch/pytorch/blob/6a3ecda5a25025d48bbc5f0215db8c338745ef79/c10/cuda/CUDAStream.cpp#L137-L138

**Comment by chang-l** [comment]
>always get the current stream from PyTorch (via TensorAdaptor maybe) instead of CUDAThreadEntry->ThreadLocal()->stream.

Are we going to always query the `current_stream` from pytorch for **EVERY** kernel calls/mem transfers? I think yes because we have no control on pytorch's stream and users may change torch current stream anywhere in their code, yet the new stream has no way propagate to DGL. This would make code a bit messy and may introduce some overhead. 

However, I agree that packing dgl stream and pytorch stream together might be more convenient and safer, from users perspective. Ideally, for now, users should set their desired stream once (for dgl and its backend), and not worry about or not encourage to 'touch' cuda-stream inside the library. How about integrate `torch.set_stream` in `dgl.set_stream`, so that (1) users only need to set stream once and only through DGL, (2) implying them not invoking `torch.set_stream` when coupled with dgl?

In terms of pytorch's implementation of backward propag., I think they use a global stream pool, defined [here](https://github.com/pytorch/pytorch/blob/ece0002c4beaebaf083dc75b7bf8ceb19edf7a0b/c10/cuda/CUDAStream.cpp#L45-L51). And they sync/record the forward streams at the entry point of autograd(backward) engine through function `InputBuffer::add`. Here are their comment descriptions: https://github.com/pytorch/pytorch/blob/ece0002c4beaebaf083dc75b7bf8ceb19edf7a0b/torch/csrc/autograd/engine.cpp#L164-L168
https://github.com/pytorch/pytorch/blob/6a3ecda5a25025d48bbc5f0215db8c338745ef79/c10/cuda/CUDAStream.h#L13-L53

**Comment by chang-l** [comment]
Maybe I can pin to torch's current stream (sth like stream = torch.current_stream()) here in the constructor of `CUDAThreadEntry`: https://github.com/dmlc/dgl/blob/2b76674002cb1df4de237024838513b6679dfbf2/src/runtime/cuda/cuda_device_api.cc#L300-L302

This way whenever it initiates a new `CUDAThreadEntry` instance at each thread, it can setup the correct stream (through torch of course).

File: src/dataloading/async_transferer.cc (hunk 3 above)
Comment (nv-dlasalle) [reply src/dataloading/async_transferer.cc:75]: Confirming that AsyncTransferer should be removed.

File: src/runtime/cuda/cuda_device_api.cc [hunk 7]
```diff
@@ -297,8 +298,13 @@ class CUDADeviceAPI final : public DeviceAPI {
 
 typedef dmlc::ThreadLocalStore<CUDAThreadEntry> CUDAThreadStore;
 
+// TODO(cliu): cuda streams should depend on the current device, therefore we should set device
+//             before setting stream.
```
Comment (chang-l) [inline src/runtime/cuda/cuda_device_api.cc:302]: Could this be error-prone of not setting device here? Now, dgl internal cuda stream is stored as a thread-local variable in `CUDAThreadEntry`. I believe it is not exposed/related to the current device in use (`DGLContext`), right?

Now, I guess it relies on whether the device has been set at beginning by pytorch.

File: src/runtime/cuda/cuda_device_api.cc [hunk 8]
```diff
@@ -297,8 +298,13 @@ class CUDADeviceAPI final : public DeviceAPI {
...
 CUDAThreadEntry::CUDAThreadEntry()
     : pool(kDLGPU, CUDADeviceAPI::Global()) {
+  TensorDispatcher* td = TensorDispatcher::Global();
+  if (td->IsAvailable())
+    stream = td->CUDAGetCurrentStream();
 }
```
Comment (chang-l) [inline src/runtime/cuda/cuda_device_api.cc:308]: Could this be error-prone of not setting device here? Now, dgl internal cuda stream is stored as a thread-local variable in CUDAThreadEntry. I believe it is not exposed/related to the current device id in use (DGLContext), right?

Now, I guess it relies on whether the device has been appropriately set at beginning by pytorch.

**Comment by chang-l** [comment]
I tested with several examples and haven't seen any "stream leakage" (pls check the updated PR description).

File: src/runtime/cuda/cuda_device_api.cc [hunk 9]
```diff
@@ -136,9 +136,10 @@ class CUDADeviceAPI final : public DeviceAPI {
                       size_t size,
                       DGLContext ctx_from,
                       DGLContext ctx_to,
-                      DGLType type_hint,
-                      DGLStreamHandle stream) final {
```
Comment (yaox12) [inline src/runtime/cuda/cuda_device_api.cc:140]: I feel for the device API, we could keep an overload version with the stream as an argument.

One difference from PyTorch is that DGL may have very complicated kernels, e.g., graph sampling, and developers may want to overlap the data transferring and computing via explicit stream control.

File: src/runtime/cuda/cuda_device_api.cc [hunk 10]
```diff
@@ -136,9 +136,10 @@ class CUDADeviceAPI final : public DeviceAPI {
...
+                      DGLType type_hint) final {
+    auto stream = static_cast<DGLStreamHandle>(CUDAThreadEntry::ThreadLocal()->stream);
     cudaStream_t cu_stream = static_cast<cudaStream_t>(stream);
```
Comment (yaox12) [inline src/runtime/cuda/cuda_device_api.cc:141]: Can be unified into one line:
```
cudaStream_t cu_stream = CUDAThreadEntry::ThreadLocal()->stream;
```

File: src/runtime/cuda/cuda_device_api.cc (hunk 8 above)
Comment (yaox12) [reply src/runtime/cuda/cuda_device_api.cc:308]: > the device has been appropriately set at beginning by pytorch.

I think it should be fine. For the single GPU case, PyTorch will use device 0 by default. While for multi-GPU DDP, users are required to set the device for each process.

File: src/runtime/cuda/cuda_device_api.cc (hunk 9 above)
Comment (chang-l) [reply src/runtime/cuda/cuda_device_api.cc:140]: I see. 
I will overload it only in cuda device API then.

File: src/runtime/cuda/cuda_device_api.cc (hunk 10 above)
Comment (chang-l) [reply src/runtime/cuda/cuda_device_api.cc:141]: lol.. Thanks. I was cpy/pasting without much thinking... Resolved as I will be outdated if I overload this function.

**Comment by chang-l** [comment]
Performed tests on hetero-rgcn and multi-gpu cases (updated PR description). No stream leakage found within dgl library calls.

**Comment by Rhett-Ying** [comment]
Commit ID: ef136364d82c0957d75a436e50ebc2d60b28f5e0

Build ID: 12

Status: ✅ CI test succeeded 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4480/12/12/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4480/12/12/logs/cireport.log)

**Comment by Rhett-Ying** [comment]
Commit ID: e9f0013ff76ac7a9d271bc356c16414a6f26c522

Build ID: 13

Status: ✅ CI test succeeded 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4480/13/13/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4480/13/13/logs/cireport.log)

**Review by yaox12 — APPROVED** [review]
LGTM.

**Comment by yaox12** [comment]
After merging this PR, I realized that to keep up with PyTorch's CUDA stream, users have to call `dgl.cuda.set_stream()` after `torch.cuda.set_stream()` every time, or use `dgl.cuda.stream()` context together with `torch.cuda.stream()`.

To simplify this, I think 
> always query the current_stream from pytorch for EVERY kernel calls/mem transfers? I think yes because we have no control on pytorch's stream and users may change torch current stream anywhere in their code, yet the new stream has no way propagate to DGL.

is acceptable, for example, PyTorch will call `c10::cuda::getCurrentCUDAStream()` everytime when it needs a stream.

To make it, I created a method `runtime::getCurrentCUDAStream()` that would query the current stream from pytorch, as a replacement of `runtime::CUDAThreadEntry::ThreadLocal()->stream` in https://github.com/dmlc/dgl/pull/4503.

**Comment by chang-l** [comment]
>I created a method runtime::getCurrentCUDAStream() that would query the current stream from pytorch, as a replacement of runtime::CUDAThreadEntry::ThreadLocal()->stream in https://github.com/dmlc/dgl/pull/4503.

I see... in this case, I wonder (1) are we going to keep `SetStream`/`CreateStream`/... methods in DGL core library? Or, they are going to in fact set/create pytorch stream instead of cuda stream? (2) if pytorch current stream is changing outside, would that introduces any sync issue to DGL core library? (e.g., when users trying to overlap some operations between dgl and pytorch and they want to call `torch.cuda.set_stream` when dgl library is running? or it should almost never happen?)

**Pull request merged**